UpdateProfile/handler.ts

```diff
@@ -0,0 +1,171 @@
+import { FiscalCodeMiddleware } from "io-functions-commons/dist/src/utils/middlewares/fiscalcode";
+import {
+  withRequestMiddlewares,
+  wrapRequestHandler
+} from "io-functions-commons/dist/src/utils/request_middleware";
+
+import { OrchestratorInput as UpsertProfileOrchestratorInput } from "../UpsertProfileOrchestrator/handler";
+import { ProfileMiddleware } from "../utils/middlewares/profile";
+import {
+  apiProfileToProfile,
+  retrievedProfileToExtendedProfile
+} from "../utils/profiles";
+
+/**
+ * Type of an UpdateProfile handler.
+ */
+type IUpdateProfileHandler = (
+  context: Context,
+  fiscalCode: FiscalCode,
+  profilePayload: ApiProfile
+) => Promise<
+  // tslint:disable-next-line: max-union-size
+  | IResponseSuccessJson<ApiProfile>
+  | IResponseErrorQuery
+  | IResponseErrorNotFound
+  | IResponseErrorConflict
+  | IResponseErrorInternal
+>;
+
+export function UpdateProfileHandler(
+  profileModel: ProfileModel
+): IUpdateProfileHandler {
+  return async (context, fiscalCode, profilePayload) => {
+    const logPrefix = `UpdateProfileHandler|FISCAL_CODE=${fiscalCode}`;
+
+    const errorOrMaybeExistingProfile = await profileModel.findOneProfileByFiscalCode(
+      fiscalCode
+    );
+
+    if (isLeft(errorOrMaybeExistingProfile)) {
+      return ResponseErrorQuery(
+        "Error trying to retrieve existing service",
+        errorOrMaybeExistingProfile.value
+      );
+    }
+
+    const maybeExistingProfile = errorOrMaybeExistingProfile.value;
+    if (isNone(maybeExistingProfile)) {
+      return ResponseErrorNotFound(
+        "Error",
+        "Could not find a profile with the provided fiscalcode"
+      );
+    }
+
+    const existingProfile = maybeExistingProfile.value;
+
+    // Verify that the client asked to update the latest version
+    if (profilePayload.version !== existingProfile.version) {
+      context.log.warn(
+        `${logPrefix}|CURRENT_VERSION=${existingProfile.version}|RESULT=CONFLICT`
+      );
+      return ResponseErrorConflict(
+        `Version ${profilePayload.version} is not the latest version.`
+      );
+    }
+
+    // Check if the email has been changed
+    const emailChanged = profilePayload.email !== existingProfile.email;
+
+    const profile = apiProfileToProfile(
+      profilePayload,
+      fiscalCode,
+      emailChanged ? false : existingProfile.isEmailValidated
+    );
+
+    const errorOrMaybeUpdatedProfile = await profileModel.update(
+      existingProfile.id,
+      existingProfile.fiscalCode,
+      p => ({
+        ...p,
+        ...profile
+      })
+    );
+
+    if (isLeft(errorOrMaybeUpdatedProfile)) {
+      context.log.error(
+        `${logPrefix}|ERROR=${errorOrMaybeUpdatedProfile.value.body}`
+      );
+      return ResponseErrorQuery(
+        "Error while updating the existing profile",
+        errorOrMaybeUpdatedProfile.value
+      );
+    }
+
+    const maybeUpdatedProfile = errorOrMaybeUpdatedProfile.value;
+
+    if (maybeUpdatedProfile.isNone()) {
```

Comment:
```suggestion
    if (isNone(maybeUpdatedProfile)) {
```

(I'd chose one between the class or functional style)